<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [1]:
import pandas as pd
import psycopg2

In [7]:
# вставьте сюда параметры подключения из юнита 1. Работа с базой данных из Python 
DBNAME = 'project_sql'
USER = 'skillfactory'
PASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt'
HOST = '84.201.134.129'
PORT = 5432


In [164]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [4]:
# текст запроса
query_3_1 = f'''
select count(*)
from public.vacancies 
'''

In [6]:
# результат запроса
df = pd.read_sql_query(query_3_1, connection)
df

C:\Users\amelnichenko\AppData\Local\Temp\ipykernel_9364\1114379196.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


count
0  49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [7]:
# текст запроса
query_3_2 = f'''
select count(*)
from public.employers 
'''

In [8]:
# результат запроса
df = pd.read_sql_query(query_3_2, connection)
df

C:\Users\amelnichenko\AppData\Local\Temp\ipykernel_9364\3062646138.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_2, connection)


count
0  23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [9]:
# текст запроса
query_3_3 = f'''
select count(*)
from public.areas 
'''

In [10]:
# результат запроса
df = pd.read_sql_query(query_3_3, connection)
df

C:\Users\amelnichenko\AppData\Local\Temp\ipykernel_9364\928149064.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_3, connection)


count
0   1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [11]:
# текст запроса
query_3_4 = f'''
select count(*)
from public.industries 
'''

In [12]:
# результат запроса
df = pd.read_sql_query(query_3_4, connection)
df

C:\Users\amelnichenko\AppData\Local\Temp\ipykernel_9364\2101691108.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_4, connection)


count
0    294

***

### Выводы по предварительному анализу данных

- количество вакансий в нашей базе (таблица vacancies) составляет:     49197 
- количество работодателей (таблица employers) составляет:             23501
- количество регионов (таблица areas) составляет:                      1362
- количество сфер деятельности в базе (таблица industries) составляет: 294



# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [20]:
# текст запроса
query_4_1 = f'''
select 
    a.name area,
    count(*) 
from 
    public.vacancies v
join public.areas a on a.id = v.area_id
group by 1
order by 2 desc
'''

In [21]:
# результат запроса
df = pd.read_sql_query(query_4_1, connection)
df

C:\Users\amelnichenko\AppData\Local\Temp\ipykernel_9364\2457173557.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_1, connection)


area  count
0             Москва   5333
1    Санкт-Петербург   2851
2              Минск   2112
3        Новосибирск   2006
4             Алматы   1892
..               ...    ...
764       Тарко-Сале      1
765    Новоаннинский      1
766            Бирск      1
767           Сасово      1
768             Анна      1

[769 rows x 2 columns]

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [22]:
# текст запроса
query_4_2 = f'''
select 
    'вакансии с указанием зп',
    count(*) 
from 
    public.vacancies v
where v.salary_from is not null or v.salary_to is not null
group by 1
'''

In [30]:
# результат запроса
df = pd.read_sql_query(query_4_2, connection)
df

C:\Users\amelnichenko\AppData\Local\Temp\ipykernel_9364\970655580.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_2, connection)


?column?  count
0  вакансии с указанием зп  24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [53]:
# текст запроса
query_4_3 = f'''
select 
    'средние значения для границ зп',
    sum(v.salary_from)/count(v.salary_from)::float salary_from,
    sum(v.salary_to)/count(v.salary_to)::float salary_to
from 
    public.vacancies v
group by 1
'''

In [54]:
# результат запроса
df = pd.read_sql_query(query_4_3, connection)
df

C:\Users\amelnichenko\AppData\Local\Temp\ipykernel_9364\2500607024.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_3, connection)


?column?   salary_from      salary_to
0  средние значения для границ зп  71064.657901  110536.741923

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [63]:
# текст запроса
query_4_4 = f'''
select 
    v.schedule,
    v.employment,
    count(*)
from 
    public.vacancies v
group by 1, 2
order by 3 DESC
'''

In [64]:
# результат запроса
df = pd.read_sql_query(query_4_4, connection)
df

C:\Users\amelnichenko\AppData\Local\Temp\ipykernel_9364\2022935265.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_4, connection)


schedule           employment  count
0        Полный день     Полная занятость  35367
1   Удаленная работа     Полная занятость   7802
2      Гибкий график     Полная занятость   1593
3   Удаленная работа  Частичная занятость   1312
4     Сменный график     Полная занятость    940
5        Полный день           Стажировка    569
6     Вахтовый метод     Полная занятость    367
7        Полный день  Частичная занятость    347
8      Гибкий график  Частичная занятость    312
9        Полный день     Проектная работа    141
10  Удаленная работа     Проектная работа    133
11     Гибкий график           Стажировка    116
12    Сменный график  Частичная занятость    101
13  Удаленная работа           Стажировка     64
14     Гибкий график     Проектная работа     18
15    Сменный график           Стажировка     12
16    Вахтовый метод     Проектная работа      2
17    Сменный график     Проектная работа      1

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [68]:
# текст запроса
query_4_5 = f'''
select 
    v.experience,
    count(*)
from 
    public.vacancies v
group by 1
order by 2 
'''

In [69]:
# результат запроса
df = pd.read_sql_query(query_4_5, connection)
df

C:\Users\amelnichenko\AppData\Local\Temp\ipykernel_9364\2316260449.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_5, connection)


experience  count
0         Более 6 лет   1337
1           Нет опыта   7197
2       От 3 до 6 лет  14511
3  От 1 года до 3 лет  26152

***

### Выводы по детальному анализу вакансий
1) Топ 5 регионов по количеству вакансий являются: Москва (5333), Санкт-Петербург (2851), Минск (2112), Новосибирск (2006), Алматы (1892)
2) Хотя бы одно из двух полей с зарплатой (от, до) заполнены у 24073 вакансий
3) Средние значения для нижней и верхней границы зарплатной вилки, округленные до целого числа: 71065р и 110537р
4) Для сочетаний типа рабочего графика (schedule) и типа трудоустройства (employment) наиболее популярными по количеству вакансий являются пары: 
    - Полный день + Полная занятость (35367 вакансий), 
    - Удаленная работа + Полная занятость (7802 вакансий), 
    - Гибкий график + Полная занятость (1593 вакансии)
5) Для требуемого опыта работы (experience) наименее популярными по количеству вакансий являются варианты:
    - Более 6 лет (1337 вакансий,
    - Нет опыта (7197) вакансий


# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [70]:
# текст запроса
query_5_1 = f'''
select 
    e.name,
    count(*)
from public.employers e
join public.vacancies v on e.id = v.employer_id 
group by 1
order by 2 DESC 
'''

In [73]:
# результат запроса
df = pd.read_sql_query(query_5_1, connection)
df

C:\Users\amelnichenko\AppData\Local\Temp\ipykernel_9364\3834399874.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_1, connection)


name  count
0                      Яндекс   1933
1                  Ростелеком    491
2                    Тинькофф    444
3                        СБЕР    428
4               Газпром нефть    331
...                       ...    ...
14761            им. Тьюринга      1
14762        LLC Novalab Tech      1
14763   Компания Мир упаковки      1
14764  Лифинцев Олег Иванович      1
14765      ТрансЭнергоХолдинг      1

[14766 rows x 2 columns]

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [104]:
# текст запроса
query_5_2 = f'''
select 
    a.name area_name,
    count(distinct v.id) vacancies_count,
    count(distinct e.id) employers_count
from public.areas a
left join public.vacancies v on a.id = v.area_id
left join public.employers e on a.id = e.area
group by a.id, a.name
order by 2 ASC, 3 DESC 
'''


In [105]:
# результат запроса
df = pd.read_sql_query(query_5_2, connection)
df

C:\Users\amelnichenko\AppData\Local\Temp\ipykernel_9364\2250345976.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_2, connection)


area_name  vacancies_count  employers_count
0                 Россия                0              410
1              Казахстан                0              207
2     Московская область                0               75
3     Краснодарский край                0               19
4     Ростовская область                0               18
...                  ...              ...              ...
1357              Алматы             1892              721
1358         Новосибирск             2006              573
1359               Минск             2112             1115
1360     Санкт-Петербург             2851             2217
1361              Москва             5333             5864

[1362 rows x 3 columns]

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [114]:
# текст запроса
query_5_3 = f'''
select 
    e.name employer_name,
    count(distinct a.id) areas_count    
from public.employers e
left join public.vacancies v on v.employer_id = e.id
left join public.areas a on a.id = v.area_id
group by e.id, e.name
order by 2 DESC 
'''

In [115]:
# результат запроса
df = pd.read_sql_query(query_5_3, connection)
df

C:\Users\amelnichenko\AppData\Local\Temp\ipykernel_9364\201667002.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_3, connection)


employer_name  areas_count
0                      Яндекс          181
1                  Ростелеком          152
2                  Спецремонт          116
3      Поляков Денис Иванович           88
4                   ООО ЕФИН            71
...                       ...          ...
23496              Но Смокинг            0
23497        МАСТЕРСКАЯ CREED            0
23498           Dzing Finance            0
23499                    КОРС            0
23500             Наука-Связь            0

[23501 rows x 2 columns]

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [129]:
# текст запроса
query_5_4 = f'''
select 
    'работодатели без указ сферы деят',
    count(distinct e.id) count    
from public.employers e
left join public.employers_industries i on e.id = i.employer_id
where i.industry_id is NULL
group by 1
'''

In [130]:
# результат запроса
df = pd.read_sql_query(query_5_4, connection)
df

C:\Users\amelnichenko\AppData\Local\Temp\ipykernel_9364\3493105328.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_4, connection)


?column?  count
0  работодатели без указ сферы деят   8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [131]:
# текст запроса
query_5_5 = f'''
select 
    e.name employer_name
from public.employers e
left join public.employers_industries i on e.id = i.employer_id
group by 1
HAVING COUNT(e.id) = 4
ORDER BY 1
'''

In [132]:
# результат запроса
df = pd.read_sql_query(query_5_5, connection)
df

C:\Users\amelnichenko\AppData\Local\Temp\ipykernel_9364\388711819.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_5, connection)


employer_name
0        101 Интернет
1            21vek.by
2                2ГИС
3                  2К
4         4 пикселя +
...               ...
1136            ЮРИОН
1137  ЮТИП Технологии
1138           ЯКласс
1139            ЯрНео
1140    Яшин&Партнёры

[1141 rows x 1 columns]

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [139]:
# текст запроса
query_5_6 = f'''
select 
    'Колич компаний со сферой деят "Разработка программного обеспечения"',
    count(distinct e.id) count
from public.employers e
join public.employers_industries ei on e.id = ei.employer_id
join public.industries i on i.id = ei.industry_id
where i.name = 'Разработка программного обеспечения'
group by 1
'''

In [140]:
# результат запроса
df = pd.read_sql_query(query_5_6, connection)
df

C:\Users\amelnichenko\AppData\Local\Temp\ipykernel_9364\821628714.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_6, connection)


?column?  count
0  Колич компаний со сферой деят "Разработка прог...   3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [156]:
# код для получения списка городов-милионников
# импортируем библиотки работы с интернет-страницами
import requests
from bs4 import BeautifulSoup

# делаем GET запрос на страницу википедии
url = 'https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8'
response = requests.get(url)

# создаем объект BeautifulSoup для парсинга HTML
soup = BeautifulSoup(response.text, 'html.parser')

# найдем таблицу с классом wikitable
table = soup.find('table', {'class': 'standard sortable'})  # standard sortable - класс таблицы (подсмотрел в HTML-коде)

# выберем из таблицы строки с нужными данными
rows = table.find_all('tr')
    
# проходим по строкам таблицы и извлекаем название города-миллионника
million_cities = []
for row in rows[1:]:  # пропускаем заголовок таблицы
    cells = row.find_all('td')
    city_name = cells[1].text.strip()  # берем 2-ю колонку таблицы - наименования городов (начиная счет с 0), первая - нумерация городов
    million_cities.append(city_name)

print(million_cities)

['Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 'Нижний Новгород', 'Красноярск', 'Челябинск', 'Самара', 'Уфа', 'Ростов-на-Дону', 'Краснодар', 'Омск', 'Воронеж', 'Пермь', 'Волгоград']


In [170]:
# текст запроса
query_5_7 = f'''
select
    a.name,
    count(distinct v.id) cnt
from public.areas a
join public.vacancies v on a.id = v.area_id
join public.employers e on e.id = v.employer_id
where (e.name = 'Яндекс') and (a.name in {tuple(million_cities)})
group by 1

UNION ALL

select
    'Total',
    count(distinct v.id) cnt
from public.areas a
join public.vacancies v on a.id = v.area_id
join public.employers e on e.id = v.employer_id
where (e.name = 'Яндекс') and (a.name in {tuple(million_cities)})
group by 1
order by 2
'''


In [171]:
# результат запроса
df = pd.read_sql_query(query_5_7, connection)
df

C:\Users\amelnichenko\AppData\Local\Temp\ipykernel_9364\1965596796.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_7, connection)


name  cnt
0              Омск   21
1         Челябинск   22
2        Красноярск   23
3         Волгоград   24
4             Пермь   25
5            Казань   25
6    Ростов-на-Дону   25
7               Уфа   26
8            Самара   26
9         Краснодар   30
10          Воронеж   32
11      Новосибирск   35
12  Нижний Новгород   36
13     Екатеринбург   39
14  Санкт-Петербург   42
15           Москва   54
16            Total  485

***

### Выводы по анализу работодателей
1. Топ 5 работодателей по количеству размещенных вакансия:
    - Яндекс 1933
    - Ростелеком 491
    - Тинькофф 444
    - СБЕР 428
    - Газпром нефть 331

2. Топ 5 регионов для которых вообше не размещены вакансии, отсортированные по количеству работодателей в них:
    - Россия	410
    - Казахстан	207
    - Московская область	75
    - Краснодарский край	19
    - Ростовская область	18
    
    Работодателям с регионами = названиям стран (Россия, Казахстан), возможно следует уточнить свое расположение для возможности более точного подбора персонала.
    Также среди работодателей этих регионов, видимо, полезно будет разместить рекламу о возможности размещения своих вакансий в системе.

3.  Топ 5 работодателей по количеству регионов, в которых они публикуют свои вакансии (по охвату регионов):
    - Яндекс	181
    - Ростелеком	152
    - Спецремонт	116
    - Поляков Денис Иванович	88
    - ООО ЕФИН	71

4. В базе вакансий имеется 8419 работодателей (из 23501), у которых не указана сфера деятельности.

5. 1140 работодателей указали 4 различные вида деятельности.

6. 3553 работодателя указали одним из видов деятельности "Разработка программного обеспечения".

7. Компания "Яндекс" разместила почти 1/4 всех вакансий (485 из 1933) в городах-миллионниках. 


# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [5]:
# текст запроса
query_6_1 = f'''
select 
    'Колич. вакансий, имеющих отношение к данным',
    count(distinct v.id) count
from public.vacancies v
where lower(v.name) LIKE '%данн%' or lower(v.name) LIKE '%data%'
group by 1
'''

In [6]:
# результат запроса
df = pd.read_sql_query(query_6_1, connection)
df

C:\Users\amelnichenko\AppData\Local\Temp\ipykernel_6156\366702258.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_1, connection)


?column?  count
0  Колич. вакансий, имеющих отношение к данным   1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [30]:
# текст запроса
query_6_2 = f'''
select 
    'Колич. подходящих вакансий для начинающего дата-сайентиста',
    count(distinct v.id) count
from public.vacancies v
where (lower(v.name) LIKE '%data scientist%' or 
       lower(v.name) LIKE '%data science%' or
       lower(v.name) LIKE '%исследователь данных%' or
       (v.name LIKE '%ML%' and not v.name LIKE '%HTML%') or
       lower(v.name) LIKE '%machine learning%' or
       lower(v.name) LIKE '%машинн%обучен%'
      ) and
      (lower(v.name) LIKE '%junior%' or
       lower(v.experience) LIKE '%нет опыта%' or
       lower(v.employment) LIKE '%стажировка%'
      )
group by 1
'''

In [31]:
# результат запроса
df = pd.read_sql_query(query_6_2, connection)
df

C:\Users\amelnichenko\AppData\Local\Temp\ipykernel_6156\2267038862.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_2, connection)


?column?  count
0  Колич. подходящих вакансий для начинающего дат...     51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [118]:
# текст запроса
query_6_3 = f'''
select 
    'Колич. вакансий для DS со знанием SQL или postgres',
    count(v.id) count
from public.vacancies v
where (lower(v.name) LIKE '%data scientist%' or 
       lower(v.name) LIKE '%data science%' or
       lower(v.name) LIKE '%исследователь данных%' or
       (v.name LIKE '%ML%' and not v.name LIKE '%HTML%') or
       lower(v.name) LIKE '%machine learning%' or
       lower(v.name) LIKE '%машинн%обучен%'
      ) and
      (lower(v.key_skills) LIKE '%sql%' or
       lower(v.key_skills) LIKE '%postgres%'
      )
group by 1
'''


In [119]:
# результат запроса
df = pd.read_sql_query(query_6_3, connection)
df

C:\Users\amelnichenko\AppData\Local\Temp\ipykernel_21968\3794299702.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_3, connection)


?column?  count
0  Колич. вакансий для DS со знанием SQL или postgres    201

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [120]:
# текст запроса
query_6_4 = f'''
select 
    'Колич. вакансий с ключевым навыком Python',
    count(*) count
from public.vacancies v
where (lower(v.name) LIKE '%data scientist%' or 
       lower(v.name) LIKE '%data science%' or
       lower(v.name) LIKE '%исследователь данных%' or
       (v.name LIKE '%ML%' and not v.name LIKE '%HTML%') or
       lower(v.name) LIKE '%machine learning%' or
       lower(v.name) LIKE '%машинн%обучен%'
      ) and
      lower(v.key_skills) LIKE '%python%'
group by 1
'''

In [121]:
# результат запроса
df = pd.read_sql_query(query_6_4, connection)
df

C:\Users\amelnichenko\AppData\Local\Temp\ipykernel_21968\4264665179.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_4, connection)


?column?  count
0  Колич. вакансий с ключевым навыком Python    351

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [110]:
# текст запроса
query_6_5 = f'''
select 
    'Средн. колич. ключевых навыков для DS',
    COUNT(v.id),
    SUM(LENGTH(v.key_skills) - LENGTH(REPLACE(v.key_skills,'\t',''))+1),
    SUM(LENGTH(v.key_skills) - LENGTH(REPLACE(v.key_skills,'\t',''))+1) / COUNT(*) ks_avg,
    SUM(LENGTH(v.key_skills) - LENGTH(REPLACE(v.key_skills,'\t',''))+1) / COUNT(*)::float ks_avg_float
from public.vacancies v
where not v.key_skills is null 
      and
      (lower(v.name) LIKE '%data scientist%' or 
       lower(v.name) LIKE '%data science%' or
       lower(v.name) LIKE '%исследователь данных%' or
       (v.name LIKE '%ML%' and not v.name LIKE '%HTML%') or
       lower(v.name) LIKE '%machine learning%' or
       lower(v.name) LIKE '%машинн%обучен%'
      ) 
group by 1
'''

In [113]:
# результат запроса
df = pd.read_sql_query(query_6_5, connection)
df

C:\Users\amelnichenko\AppData\Local\Temp\ipykernel_21968\3654210517.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_5, connection)


?column?  count   sum  ks_avg  ks_avg_float
0  Средн. колич. ключевых навыков для DS    431  2761       6      6.406032

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [156]:
# текст запроса
query_6_6 = f'''
select 
    v.experience,
    AVG( 
    COALESCE((v.salary_to + v.salary_from)/2, COALESCE(v.salary_to, 0) + COALESCE(v.salary_from, 0))
    ) 
    AS AvgVal
from public.vacancies v
where (lower(v.name) LIKE '%data scientist%' or 
       lower(v.name) LIKE '%data science%' or
       lower(v.name) LIKE '%исследователь данных%' or
       (v.name LIKE '%ML%' and not v.name LIKE '%HTML%') or
       lower(v.name) LIKE '%machine learning%' or
       lower(v.name) LIKE '%машинн%обучен%'
      ) and
      (v.salary_to IS NOT NULL or v.salary_from IS NOT NULL)
group by 1
'''

In [157]:
# результат запроса
df = pd.read_sql_query(query_6_6, connection)
df

C:\Users\amelnichenko\AppData\Local\Temp\ipykernel_21968\519194770.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_6, connection)


experience         avgval
0           Нет опыта   74642.857143
1  От 1 года до 3 лет  139674.750000
2       От 3 до 6 лет  243114.666667

***

### Выводы по предметному анализу
1) Количество вакансий, имеющих отношение к данным, = 1771 (из 49197) - т.е. около 3,5%. 
Количество подходящих вакансий для дата-сайентиста = 480 (менее 1%), а для начинающего дата-сайентиста = 51 (0,1%)
Это говорит, во-первых, о достаточно высокой уникальности специальности дата-сайентиста, а во-вторых, о том, что востребованность опытных дата-сайентистов примерно в 10 раз выше, чем начинающих. Кроме того, средняя зарплата по вакансиям для дата-сайентиста быстро растет с увеличением опыта (74тр для начинающего без опыта и 243тр при наличии опыта от 3 до 6 лет). Это говорит о высоких перспективах для карьерного и зарплатного роста.
2) Однако, профессия дата-сайентиста требует широких ключевых навыков в различных направлениях, так, в вакансиях для DS в среднем указывают 6,4 ключевых навыка. среди которых практически всегда присутствуют знания Python, SQL и Postgres. Следовательно, компетенции в этих областях крайне важны для успеха дата-сайентиста.

# Общий вывод по проекту

In [ ]:
# подведем итог исследования, обобщите выводы
# здесь можно (это будет плюсом) провести дополнительные исследования данных, сделать прогнозы, продумать варианты продолжения исследования

### В результате работы по проекту были проведены следующие исследования:
1) ПРЕДВАРИТЕЛЬНЫЙ АНАЛИЗ ДАННЫХ - определение основных количественных параметров базы данных (количество вакансий в нашей базе, количество работодателей, количество регионов, количество сфер деятельности).

2) ДЕТАЛЬНЫЙ АНАЛИЗ ВАКАНСИЙ (топ 5 регионов по количеству вакансий, количество вакансий с заполненным уровнем ЗП, средние значения для нижней и верхней границ ЗП, наиболее популярные сочетания типа рабочего графика и типа трудоустройства, наименее популярные варианты опыта работы). 

3) АНАЛИЗ РАБОТОДАТЕЛЕЙ (топ 5 работодателей по количеству размещенных вакансия, топ 5 регионов для которых вообше не размещены вакансии, топ 5 работодателей по охвату регионов, количество работодателей, у которых не указана сфера деятельности, количество работодателей, указавших 4 различных вида деятельности, количество работодателей, указавших 
одним из видов деятельности "Разработка программного обеспечения", анализ размещения работодателями вакансий 
в городах-миллионниках). 

4) ПРЕДМЕТНЫЙ АНАЛИЗ - анализ показателей непосредственно для вакансии дата-сайентиста (количество вакансий, в том числе для уровня junior, уровень средней ЗП от требуемого опыта, требования к ключевым навыкам).

#### В итоге исследования мы достигли следующих результатов:

- поняли общие параметры рынка трудоустройства;

- оценили условия и возможности по трудоустройству для потенциальных кандидатов, в том числе, для дата-сайентистов;

- выявили недостатки информации в базе (вакансии с не заполненной ЗП, работодатели с не указанной сферой деятельности, регионы с неразмещенными вакансиями).

#### Результаты исследование полезны, как для кандидатов, так и для работодателей и организаторов сбора информации по вакансиям (в части размещения подсказок и рекомендаций по заполнению информации).


#### В качестве дополнительного исследования данных можно предложить:

- ДОП_1. Исследование среднего уровня зарплаты дата-сайентиста в разрезе Регион + Работодатель с сортировкой в порядке уменьшения средней ЗП.

- ДОП_2. Исследование имеющихся вариантов подработки для DS (вариант трудоустройства = Частичная занятость) с выводом Региона, Работодателя и зарплатной вилки.

In [178]:
# ДОП_1. 
# Исследование среднего уровня зарплаты дата-сайентиста в разрезе Регион + Работодатель 
# с сортировкой в порядке уменьшения средней ЗП
query_7_1 = f'''
select 
    a.name Регион,
    e.name Работодатель,
    AVG( 
    COALESCE((v.salary_to + v.salary_from)/2, COALESCE(v.salary_to, 0) + COALESCE(v.salary_from, 0))
    ) 
    AS Средняя_ЗП
from public.vacancies v
left join public.employers e on v.employer_id = e.id
left join public.areas a on v.area_id = a.id
where (lower(v.name) LIKE '%data scientist%' or 
       lower(v.name) LIKE '%data science%' or
       lower(v.name) LIKE '%исследователь данных%' or
       (v.name LIKE '%ML%' and not v.name LIKE '%HTML%') or
       lower(v.name) LIKE '%machine learning%' or
       lower(v.name) LIKE '%машинн%обучен%'
      ) and
      (v.salary_to IS NOT NULL or v.salary_from IS NOT NULL)
group by 1, 2
order by 3 DESC
'''
df = pd.read_sql_query(query_7_1, connection)
df

C:\Users\amelnichenko\AppData\Local\Temp\ipykernel_21968\3074362068.py:25: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_7_1, connection)


Регион                                        Работодатель  \
0            Москва                                                Итон   
1            Москва                                          AppQuantum   
2            Москва                                           JSA Group   
3       Новосибирск                                                Дром   
4   Санкт-Петербург                                             OVISION   
5   Санкт-Петербург                                           JSA Group   
6              Кипр                                             Semrush   
7            Москва                                          СберЛизинг   
8            Москва                                           КИБЕР-РОМ   
9            Москва                                              Dbrain   
10           Москва                                        Кловер Групп   
11          Армения                                            Andersen   
12           Москва                         Freedom International Group   
13           Москва                                        Spice Agency   
14           Турция                                      IntelligiChain   
15           Сербия                                      IntelligiChain   
16       Черногория                                      IntelligiChain   
17     Екатеринбург                                           Сима-ленд   
18           Москва                                              Bip.ru   
19           Москва                                              Телега   
20           Москва                                          АЙТИ.СПЕЙС   
21           Москва                                          DIGINETICA   
22         Белгород                                  Зеленая долина, ГК   
23           Москва                             Алгоритмические Системы   
24           Москва                                        УЦ Новый шаг   
25           Москва                                          JuicyScore   
26  Санкт-Петербург                                        Hype Auditor   
27  Санкт-Петербург                                               НИИАС   
28           Москва             Центр хранения и анализа больших данных   
29           Рязань                                          MNOGOGAMES   
30           Москва                            Интеллектуальные решения   
31           Москва                                        НСК ЭкоСфера   
32           Москва                                      РТК Автоматика   
33           Москва                                   Азимут Технолоджи   
34           Москва                                         РКК Энергия   
35          Королев                                         РКК Энергия   
36           Рязань                                  Программный регион   
37           Алматы                                        Kolesa Group   
38           Москва                                             Лидстех   
39      Новосибирск                            Igor Polyakov Consulting   
40  Санкт-Петербург                                    Университет ИТМО   
41     Екатеринбург                          Сталепромышленная компания   
42           Москва                               ВижнЛабс (VisionLabs)   
43  Санкт-Петербург                                            ННФормат   
44              Уфа                                            Homework   
45            Пермь                                            Homework   
46      Владивосток                                   DNS Головной офис   
47      Новосибирск                                             Апэрбот   
48           Москва                                   Cpaaggregator.com   
49           Москва                                   Строительный Двор   
50           Алматы                                 Cybernet (Кибернет)   
51          Барнаул                                             Энтерра   
52            Пермь                   

In [170]:

# ДОП_2.
# Исследование имеющихся вариантов подработки для DS (вариант трудоустройства = Частичная занятость)
# с выводом региона, работодателя и зарплатной вилки
query_7_2 = f'''
select
    a.name Регион,
    e.name Работодатель,
    v.salary_from ЗП_от,
    v.salary_to ЗП_до
from public.vacancies v
left join public.employers e on v.employer_id = e.id
left join public.areas a on v.area_id = a.id
where (lower(v.name) LIKE '%data scientist%' or 
       lower(v.name) LIKE '%data science%' or
       lower(v.name) LIKE '%исследователь данных%' or
       (v.name LIKE '%ML%' and not v.name LIKE '%HTML%') or
       lower(v.name) LIKE '%machine learning%' or
       lower(v.name) LIKE '%машинн%обучен%'
       ) 
       and
       v.employment = 'Частичная занятость'
order by 1, 2
'''
df = pd.read_sql_query(query_7_2, connection)
df

C:\Users\amelnichenko\AppData\Local\Temp\ipykernel_21968\3956880336.py:23: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_7_2, connection)


Регион  \
0            Минск   
1           Москва   
2           Москва   
3           Москва   
4      Новосибирск   
5      Новосибирск   
6  Санкт-Петербург   
7            Томск   

                                                   Работодатель     ЗП_от  \
0  Представительство ООО «Квалитет» (Российская Федерация) в РБ       NaN   
1                                             Строительный Двор   45000.0   
2                                                        Яндекс       NaN   
3                                              Яндекс Практикум       NaN   
4                                                        Comexp       NaN   
5                                      Igor Polyakov Consulting  119000.0   
6                                                        Comexp       NaN   
7                                                        Comexp       NaN   

      ЗП_до  
0       NaN  
1  100000.0  
2       NaN  
3       NaN  
4       NaN  
5       NaN  
6       NaN  
7       NaN